# Introduction - Oracle Version 2

Notebook to load **bug reports**, **test cases** datasets and the **feature_matrixes** from the expert and volunteers responses given in the PyBossa applications, and create from them the **oracle** dataset. 

In this notebook we create a version of the oracle based on the results obtained from the empirical study made with volunteers and the PyBossa application. The relationship between bug reports and test cases is established through the firefox features shared between both artifacts.

Is expected that this oracle be more precise than the version created on the previous notebook (__oracle_v1__), once the trace links are created based on the existing relationship between the bug report and a given Firefox Feature.

## Load Libraries and Data

In [10]:
import pandas as pd
import numpy as np
from sklearn.externals.joblib import Parallel, delayed
from tqdm import tqdm

from utils import aux_functions

testcases = pd.read_csv('../data/mozilla_firefox_v2/firefoxDataset/docs_english/TC/testcases_final.csv')
print('Test Cases Shape: {}'.format(testcases.shape))

bugreports = pd.read_csv('../data/mozilla_firefox_v2/firefoxDataset/docs_english/BR/bugreports_final.csv')
print('Bug Reports shape: {}'.format(bugreports.shape))

expert_matrix = pd.read_csv('../data/mozilla_firefox_v2/firefoxDataset/br_feat_recovery_empirical_study/pybossa-apps/recover_taskruns/br_2_feature_matrix_expert.csv')
expert_matrix.set_index('bug_number', inplace=True)
print('Expert Matrix shape: {}'.format(expert_matrix.shape))

volunteers_matrix = pd.read_csv('../data/mozilla_firefox_v2/firefoxDataset/br_feat_recovery_empirical_study/pybossa-apps/recover_taskruns/br_2_feature_matrix_volunteers.csv')
volunteers_matrix.set_index('bug_number', inplace=True)
print('Volunteers Matrix shape: {}'.format(volunteers_matrix.shape))

features = pd.read_csv('../data/mozilla_firefox_v2/firefoxDataset/docs_english/Features/features.csv')
print('Features shape: {}'.format(features.shape))

Test Cases Shape: (207, 12)
Bug Reports shape: (35336, 18)
Expert Matrix shape: (93, 21)
Volunteers Matrix shape: (30, 21)
Features shape: (21, 5)


## EDA - Exploratory Data Analysis

In [11]:
testcases.head()

TC_Number   TestDay  Feature_ID  \
0          1  20181221          20   
1          2  20181221          20   
2          3  20181221          20   
3          4  20181221          20   
4          5  20181221          20   

                                Firefox_Feature  \
0  <notificationbox> and <notification> changes   
1  <notificationbox> and <notification> changes   
2  <notificationbox> and <notification> changes   
3  <notificationbox> and <notification> changes   
4  <notificationbox> and <notification> changes   

                                      Gen_Title  Crt_Nr  \
0  <notificationbox> and <notification> changes       1   
1  <notificationbox> and <notification> changes       2   
2  <notificationbox> and <notification> changes       3   
3  <notificationbox> and <notification> changes       4   
4  <notificationbox> and <notification> changes       5   

                                               Title Preconditions  \
0                         Notification - Popup Block           NaN   
1                        Notification - Process Hang           NaN   
2            Verify Notifications appear in RTL Mode           NaN   
3  Verify Notifications appear in High Contrast M...           NaN   
4  Verify notifications react to differing Zoom l...           NaN   

                                               Steps  \
0  1. Launch Firefox\n2. Navigate to http://www.p...   
1  1. Launch Firefox\n2. In the URL bar, navigate...   
2  1. Launch Firefox\n2. In about:config, change ...   
3  1. While the browser is in High Contrast Mode,...   
4  1. While the browser is in High Contrast Mode,...   

                                     Expected_Result   tc_name  \
0  1. Firefox is successfully launched\n9. The al...  TC_1_TRG   
1  1. Firefox is successfully launched\n2. Firefo...  TC_2_TRG   
2  1. Firefox is successfully launched\n2.The for...  TC_3_TRG   
3  1. Firefox has been launched.\n2. Firefox begi...  TC_4_TRG   
4  1. Firefox has been launched.\n2. Firefox begi...  TC_5_TRG   

                                             tc_desc  
0  1 20181221 20 <notificationbox> and <notificat...  
1  2 20181221 20 <notificationbox> and <notificat...  
2  3 20181221 20 <notificationbox> and <notificat...  
3  4 20181221 20 <notificationbox> and <notificat...  
4  5 20181221 20 <notificationbox> and <notificat...

In [12]:
bugreports.head()

Bug_Number                                            Summary Platform  \
0      506297  Livemarks with null site/feed uris cause sync ...      All   
1      506338     Enhance Crash Recovery to better help the user      All   
2      506507  Dragging multiple bookmarks in the bookmarks s...      x86   
3      506550      Unreliable Back Button navigating nytimes.com      x86   
4      506575  ALT + F4 when dropdown of autocomplete is open...      x86   

                 Component      Version         Creation_Time  \
0                     Sync  unspecified  2009-07-24T17:08:43Z   
1          Session Restore        Trunk  2009-07-24T19:17:21Z   
2      Bookmarks & History        Trunk  2009-07-26T06:16:02Z   
3  Extension Compatibility   3.5 Branch  2009-07-26T16:12:10Z   
4              Address Bar   3.5 Branch  2009-07-26T20:14:54Z   

                                       Whiteboard QA_Whiteboard  \
0                                             NaN           NaN   
1                  [crashkill][crashkill-metrics]           NaN   
2                                             NaN           NaN   
3  [caused by adblock plus][platform-rel-NYTimes]           NaN   
4                                             NaN           NaN   

                                  First_Comment_Text  \
0  2009-07-24 09:54:28     FaultTolerance       D...   
1  When our users crash they are pretty much in t...   
2  User-Agent:       Mozilla/5.0 (Windows; U; Win...   
3  User-Agent:       Mozilla/5.0 (Windows; U; Win...   
4  Pressing ALT + F4 when the autocomplete dropdo...   

  First_Comment_Creation_Time    Status  Product Priority  Resolution  \
0        2009-07-24T17:08:43Z  RESOLVED  Firefox       --       FIXED   
1        2009-07-24T19:17:21Z       NEW  Firefox       --         NaN   
2        2009-07-26T06:16:02Z  RESOLVED  Firefox       --  WORKSFORME   
3        2009-07-26T16:12:10Z  RESOLVED  Firefox       --       FIXED   
4        2009-07-26T20:14:54Z       NEW  Firefox       P5         NaN   

      Severity Is_Confirmed        br_name  \
0       normal         True  BR_506297_SRC   
1  enhancement         True  BR_506338_SRC   
2       normal         True  BR_506507_SRC   
3       normal        False  BR_506550_SRC   
4       normal         True  BR_506575_SRC   

                                             br_desc  
0  506297 Livemarks with null site/feed uris caus...  
1  506338 Enhance Crash Recovery to better help t...  
2  506507 Dragging multiple bookmarks in the book...  
3  506550 Unreliable Back Button navigating nytim...  
4  506575 ALT + F4 when dropdown of autocomplete ...

In [13]:
features.head()

Feature_Number        Firefox_Version        Firefox_Feature  \
0               1  48 Branch + 50 Branch        New Awesome Bar   
1               2              48 Branch     Windows Child Mode   
2               3              48 Branch  APZ - Async Scrolling   
3               4              49 Branch  Browser Customization   
4               5              49 Branch             PDF Viewer   

                                 Feature_Description  \
0  The Firefox address bar displays a page's web ...   
1  Child mode is a feature of Windows that allows...   
2  The Async Pan/Zoom module (APZ) is a platform ...   
3  . Install and Customize Firefox Themes. \n. Cu...   
4       Zoom in, Zoom out, Print and Save PDF Files.   

                                           Reference  
0  https://support.mozilla.org/en-US/kb/awesome-b...  
1     https://wiki.mozilla.org/QA/Windows_Child_Mode  
2          https://wiki.mozilla.org/Platform/GFX/APZ  
3  https://support.mozilla.org/en-US/kb/use-theme...  
4  https://support.mozilla.org/en-US/kb/view-pdf-...

In [14]:
expert_matrix.head()

new_awesome_bar  windows_child_mode  apz_async_scrolling  \
bug_number                                                             
1181835                   0                   0                    0   
1248267                   0                   0                    0   
1248268                   0                   0                    0   
1257087                   1                   0                    0   
1264988                   0                   0                    0   

            browser_customization  pdf_viewer  context_menu  w10_comp  \
bug_number                                                              
1181835                         0           0             1         0   
1248267                         0           0             0         0   
1248268                         0           0             0         0   
1257087                         0           0             0         0   
1264988                         0           0             0         0   

            tts_in_desktop  tts_in_rm  webgl_comp        ...         \
bug_number                                               ...          
1181835                  0          0           0        ...          
1248267                  0          0           0        ...          
1248268                  0          0           0        ...          
1257087                  0          0           0        ...          
1264988                  0          0           0        ...          

            pointer_lock_api  webm_eme  zoom_indicator  downloads_dropmaker  \
bug_number                                                                    
1181835                    0         0               0                    0   
1248267                    0         0               0                    0   
1248268                    0         0               0                    0   
1257087                    0         0               0                    0   
1264988                    0         0               0                    0   

            webgl2  flac_support  indicator_device_perm  flash_support  \
bug_number                                                               
1181835          0             0                      0              0   
1248267          0             0                      0              0   
1248268          0             0                      0              0   
1257087          0             0                      0              0   
1264988          0             0                      0              0   

            notificationbox  update_directory  
bug_number                                     
1181835                   0                 0  
1248267                   0                 0  
1248268                   0                 0  
1257087                   0                 0  
1264988                   0                 0  

[5 rows x 21 columns]

In [15]:
volunteers_matrix.head()

new_awesome_bar  windows_child_mode  apz_async_scrolling  \
bug_number                                                             
1181835                   0                   0                    0   
1248267                   0                   0                    0   
1248268                   0                   0                    0   
1257087                   1                   0                    0   
1264988                   0                   0                    0   

            browser_customization  pdf_viewer  context_menu  w10_comp  \
bug_number                                                              
1181835                         0           0             1         0   
1248267                         1           0             0         0   
1248268                         0           0             0         0   
1257087                         0           0             0         0   
1264988                         0           0             0         0   

            tts_in_desktop  tts_in_rm  webgl_comp        ...         \
bug_number                                               ...          
1181835                  0          0           0        ...          
1248267                  0          0           0        ...          
1248268                  0          0           0        ...          
1257087                  0          0           0        ...          
1264988                  0          0           0        ...          

            pointer_lock_api  webm_eme  zoom_indicator  downloads_dropmaker  \
bug_number                                                                    
1181835                    0         0               0                    0   
1248267                    0         0               0                    0   
1248268                    0         0               0                    0   
1257087                    0         0               0                    0   
1264988                    0         0               0                    0   

            webgl2  flac_support  indicator_device_perm  flash_support  \
bug_number                                                               
1181835          0             0                      0              0   
1248267          0             0                      0              0   
1248268          0             0                      0              0   
1257087          0             0                      0              0   
1264988          0             0                      0              0   

            notificationbox  update_directory  
bug_number                                     
1181835                   0                 0  
1248267                   0                 0  
1248268                   0                 0  
1257087                   0                 0  
1264988                   0                 0  

[5 rows x 21 columns]

### Bug Reports Selection

In [16]:
brs_versions = ['48 Branch', '49 Branch', '50 Branch', '51 Branch']
brs_status = ['RESOLVED','VERIFIED']
brs_priority = ['P1', 'P2', 'P3']
brs_resolutions = ['FIXED']
brs_severities = ['major', 'normal', 'blocker', 'critical']
brs_isconfirmed = [True]
selected_bugs = bugreports[(bugreports.Version.isin(brs_versions)) &
                             (bugreports.Status.isin(brs_status)) &
                             (bugreports.Priority.isin(brs_priority)) &
                             (bugreports.Resolution.isin(brs_resolutions)) &
                             (bugreports.Severity.isin(brs_severities)) &
                             (bugreports.Is_Confirmed.isin(brs_isconfirmed))
                            ]
print(selected_bugs.shape)

(93, 18)


## Oracle

#### Add Feature_Name column to features dataset

In [17]:
features['Feature_Name'] = ['new_awesome_bar',
            'windows_child_mode',
            'apz_async_scrolling',
            'browser_customization',
            'pdf_viewer',
            'context_menu',
            'w10_comp', 
            'tts_in_desktop', 
            'tts_in_rm', 
            'webgl_comp',
            'video_and_canvas_render', 
            'pointer_lock_api',
            'webm_eme', 
            'zoom_indicator',
            'downloads_dropmaker',
            'webgl2', 
            'flac_support', 
            'indicator_device_perm',
            'flash_support',  
            'notificationbox',          
            'update_directory']
features.head(22)

Feature_Number        Firefox_Version  \
0                1  48 Branch + 50 Branch   
1                2              48 Branch   
2                3              48 Branch   
3                4              49 Branch   
4                5              49 Branch   
5                6              49 Branch   
6                7              49 Branch   
7                8              49 Branch   
8                9              49 Branch   
9               10              49 Branch   
10              11              49 Branch   
11              12              50 Branch   
12              13              50 Branch   
13              14              51 Branch   
14              15              51 Branch   
15              16              51 Branch   
16              17              51 Branch   
17              18              51 Branch   
18              19              51 Branch   
19              20              65 Branch   
20              21              65 Branch   

                                 Firefox_Feature  \
0                                New Awesome Bar   
1                             Windows Child Mode   
2                          APZ - Async Scrolling   
3                          Browser Customization   
4                                     PDF Viewer   
5                                   Context Menu   
6                       Windows 10 Compatibility   
7                      Text to Speech on Desktop   
8                  Text to Speech in Reader Mode   
9                            WebGL Compatibility   
10                Video and Canvas Renderization   
11                              Pointer Lock API   
12                 WebM EME support for Widevine   
13                               Zoom Indicator    
14                           Downloads Dropmaker   
15                                        WebGL2   
16                                  FLAC support   
17              Indicator for device permissions   
18                                 Flash support   
19  <notificationbox> and <notification> changes   
20                              Update Directory   

                                  Feature_Description  \
0   The Firefox address bar displays a page's web ...   
1   Child mode is a feature of Windows that allows...   
2   The Async Pan/Zoom module (APZ) is a platform ...   
3   . Install and Customize Firefox Themes. \n. Cu...   
4        Zoom in, Zoom out, Print and Save PDF Files.   
5   The Context Menu is the menu that appears when...   
6   . Users can browse through different sites, \n...   
7   Extension of Firefox to read out loud the HTML...   
8   Extension of Firefox to read out loud the cont...   
9   Feature of Firefox implemted based on WebGL li...   
10  Feature of Firefox implemented for rendering o...   
11  The Pointer Lock API (formerly called Mouse Lo...   
12  Encrypted Media Extensions (EME) is a JavaScri...   
13  The Zoom Indicator shows to the user the amoun...   
14  The download dropmaker displays the amount of ...   
15  Incorporation of WebGL2 library to Firefox Bro...   
16  FLAC stands for Free Lossless Audio Codec, an ...   
17  The indicator for device permissions displays ...   
18  The Adobe Flash Player is commonly used on the...   
19  The notificationbox element is used to display...   
20  The Update Directory feature is implemented to...   

                                            Reference             Feature_Name  
0   https://support.mozilla.org/en-US/kb/awesome-b...          new_awesome_bar  
1      https://wiki.mozilla.org/QA/Windows_Child_Mode       windows_child_mode  
2           https://wiki.mozilla.org/Platform/GFX/APZ      apz_async_scrolling  
3   https://support.mozilla.org/en-US/kb/use-theme...    browser_customization  
4   https://support.mozilla.org/en-US/kb/view-pdf-...               pdf_viewer  
5   https://www.paessler.com/manuals/paessler_site...             context_menu  
6   https://support.mozilla.org/en-US/kb/how-d

In [18]:
display(expert_matrix.head())
expert_matrix.iloc[0,5]

new_awesome_bar  windows_child_mode  apz_async_scrolling  \
bug_number                                                             
1181835                   0                   0                    0   
1248267                   0                   0                    0   
1248268                   0                   0                    0   
1257087                   1                   0                    0   
1264988                   0                   0                    0   

            browser_customization  pdf_viewer  context_menu  w10_comp  \
bug_number                                                              
1181835                         0           0             1         0   
1248267                         0           0             0         0   
1248268                         0           0             0         0   
1257087                         0           0             0         0   
1264988                         0           0             0         0   

            tts_in_desktop  tts_in_rm  webgl_comp        ...         \
bug_number                                               ...          
1181835                  0          0           0        ...          
1248267                  0          0           0        ...          
1248268                  0          0           0        ...          
1257087                  0          0           0        ...          
1264988                  0          0           0        ...          

            pointer_lock_api  webm_eme  zoom_indicator  downloads_dropmaker  \
bug_number                                                                    
1181835                    0         0               0                    0   
1248267                    0         0               0                    0   
1248268                    0         0               0                    0   
1257087                    0         0               0                    0   
1264988                    0         0               0                    0   

            webgl2  flac_support  indicator_device_perm  flash_support  \
bug_number                                                               
1181835          0             0                      0              0   
1248267          0             0                      0              0   
1248268          0             0                      0              0   
1257087          0             0                      0              0   
1264988          0             0                      0              0   

            notificationbox  update_directory  
bug_number                                     
1181835                   0                 0  
1248267                   0                 0  
1248268                   0                 0  
1257087                   0                 0  
1264988                   0                 0  

[5 rows x 21 columns]

1

In [19]:
def get_features_from_expert_matrix(br_id):
    features_ids = ""
    for col in expert_matrix.columns:
        if expert_matrix.at[br_id, col] == 1:
            if features_ids == "":
                features_ids = str(expert_matrix.columns.get_loc(col) + 1)
            else:
                features_ids = features_ids + " " + str(expert_matrix.columns.get_loc(col) + 1)
    return features_ids

selected_bugs['Features_IDs'] = ""
for br_id in expert_matrix.index:
    for idx2, br in selected_bugs.iterrows():
        if br.Bug_Number == br_id:
            selected_bugs.at[idx2, 'Features_IDs'] = get_features_from_expert_matrix(br_id)

selected_bugs[['Bug_Number','Features_IDs']].head()

/home/guilherme/anaconda3/envs/trace-link-recovery-study/lib/python3.6/site-packages/ipykernel_launcher.py:11: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  # This is added back by InteractiveShellApp.init_path()


Bug_Number Features_IDs
9192      1181835            6
10621     1248267             
10622     1248268             
10904     1257087            1
11150     1264988

In [20]:
selected_bugs.head()

Bug_Number                                            Summary  \
9192      1181835  Provide a UI for migrating users' add-ons to w...   
10621     1248267  Right click on bookmark item of "Recently Book...   
10622     1248268            Unable to disable "Recently bookmarked"   
10904     1257087  Middle mouse click on history item would not open   
11150     1264988  Scrollbar appears for a moment in the new Awes...   

          Platform                Component    Version         Creation_Time  \
9192   Unspecified  Extension Compatibility  49 Branch  2015-07-08T23:01:45Z   
10621  Unspecified      Bookmarks & History  48 Branch  2016-02-14T17:45:54Z   
10622          All      Bookmarks & History  48 Branch  2016-02-14T17:54:44Z   
10904  Unspecified      Bookmarks & History  48 Branch  2016-03-16T05:13:47Z   
11150          All              Address Bar  48 Branch  2016-04-15T15:17:33Z   

                            Whiteboard QA_Whiteboard  \
9192                      [UX] triaged           NaN   
10621                              NaN           NaN   
10622                              NaN           NaN   
10904                              NaN           NaN   
11150  [fxsearch] [photon-performance]           NaN   

                                      First_Comment_Text  \
9192   We're still not exactly sure how this would wo...   
10621  Steps To Reproduce: 1. Open Bookmarks menu 2. ...   
10622  Created attachment 8719295 Firefox Nightly 47_...   
10904  [Tracking Requested - why for this release]: r...   
11150  Created attachment 8741829 Bug.mov  User Agent...   

      First_Comment_Creation_Time    Status  Product Priority Resolution  \
9192         2015-07-08T23:01:45Z  RESOLVED  Firefox       P2      FIXED   
10621        2016-02-14T17:45:54Z  VERIFIED  Firefox       P3      FIXED   
10622        2016-02-14T17:54:44Z  VERIFIED  Firefox       P3      FIXED   
10904        2016-03-16T05:13:47Z  VERIFIED  Firefox       P2      FIXED   
11150        2016-04-15T15:17:33Z  VERIFIED  Firefox       P1      FIXED   

      Severity Is_Confirmed         br_name  \
9192    normal         True  BR_1181835_SRC   
10621   normal         True  BR_1248267_SRC   
10622    major         True  BR_1248268_SRC   
10904   normal         True  BR_1257087_SRC   
11150   normal         True  BR_1264988_SRC   

                                                 br_desc Features_IDs  
9192   1181835 Provide a UI for migrating users' add-...            6  
10621  1248267 Right click on bookmark item of "Recen...               
10622  1248268 Unable to disable "Recently bookmarked...               
10904  1257087 Middle mouse click on history item wou...            1  
11150  1264988 Scrollbar appears for a moment in the ...

In [21]:
def check_link_condition(br, tc):
    if str(tc.Feature_ID) in br.Features_IDs.split(" "):
            return True
    return False

oracle_df = pd.DataFrame(columns=selected_bugs.br_name, index=testcases.tc_name, data=np.zeros(shape=(len(testcases),len(selected_bugs))), dtype='int8')
for idx_1,br in tqdm(selected_bugs.iterrows()):
    for idx_2,tc in testcases.iterrows():
        if check_link_condition(br, tc):
            oracle_df.at[tc.tc_name, br.br_name] = 1
        else:
            oracle_df.at[tc.tc_name, br.br_name] = 0


93it [00:01, 47.53it/s]


In [25]:
print('Oracle shape: {}\n'.format(oracle_df.shape))
display(oracle_df.head())

Oracle shape: (207, 93)



br_name   BR_1181835_SRC  BR_1248267_SRC  BR_1248268_SRC  BR_1257087_SRC  \
tc_name                                                                    
TC_1_TRG               0               0               0               0   
TC_2_TRG               0               0               0               0   
TC_3_TRG               0               0               0               0   
TC_4_TRG               0               0               0               0   
TC_5_TRG               0               0               0               0   

br_name   BR_1264988_SRC  BR_1267480_SRC  BR_1267501_SRC  BR_1269348_SRC  \
tc_name                                                                    
TC_1_TRG               0               0               0               0   
TC_2_TRG               0               0               0               0   
TC_3_TRG               0               0               0               0   
TC_4_TRG               0               0               0               0   
TC_5_TRG               0               0               0               0   

br_name   BR_1269485_SRC  BR_1270274_SRC       ...        BR_1352539_SRC  \
tc_name                                        ...                         
TC_1_TRG               0               0       ...                     0   
TC_2_TRG               0               0       ...                     0   
TC_3_TRG               0               0       ...                     0   
TC_4_TRG               0               0       ...                     0   
TC_5_TRG               0               0       ...                     0   

br_name   BR_1353831_SRC  BR_1357085_SRC  BR_1357458_SRC  BR_1365887_SRC  \
tc_name                                                                    
TC_1_TRG               0               0               0               1   
TC_2_TRG               0               0               0               1   
TC_3_TRG               0               0               0               1   
TC_4_TRG               0               0               0               1   
TC_5_TRG               0               0               0               1   

br_name   BR_1408361_SRC  BR_1430603_SRC  BR_1432915_SRC  BR_1449700_SRC  \
tc_name                                                                    
TC_1_TRG               0               0               0               0   
TC_2_TRG               0               0               0               0   
TC_3_TRG               0               0               0               0   
TC_4_TRG               0               0               0               0   
TC_5_TRG               0               0               0               0   

br_name   BR_1451475_SRC  
tc_name                   
TC_1_TRG               0  
TC_2_TRG               0  
TC_3_TRG               0  
TC_4_TRG               0  
TC_5_TRG               0  

[5 rows x 93 columns]

### Save Datasets
New Oracle and Selected Bug Reports Dataset
Dimensions: 
 * 207 x 93
 * 93 x 19

In [84]:
oracle_df.to_csv('../data/mozilla_firefox_v2/firefoxDataset/oracle/output/trace_matrix_final_emp_study.csv')
selected_bugs.to_csv('../data/mozilla_firefox_v2/firefoxDataset/docs_english/BR/selected_bugreports_final.csv', index=False)

## -----------------------------------------

#### Checking Values [0]

#### Analyze Entire Oracle Created

In [85]:
oo_df_full = pd.read_csv('../data/mozilla_firefox_v2/firefoxDataset/oracle/output/trace_matrix_final_emp_study.csv')
oo_df_full.set_index('tc_name', inplace=True)

assert(oo_df_full.loc['TC_13_TRG', 'BR_1257087_SRC'] == 1)
assert(oo_df_full.loc['TC_14_TRG', 'BR_1271607_SRC'] == 1)
assert(oo_df_full.loc['TC_15_TRG', 'BR_1276120_SRC'] == 1)

### -----

#### Checking Values [1]
Amount Positive and Negative Links

In [86]:
positive_links = 0
negative_links = 0
for idx,row in oo_df_full.iterrows():
    for col in oo_df_full.columns:
        if row[col] == 1:
            positive_links = positive_links + 1
        else:
            negative_links = negative_links + 1

print("Positive Links Amount: {}".format(positive_links))
print("Negative Links Amount: {}".format(negative_links))

Positive Links Amount: 915
Negative Links Amount: 18336


##### Expected Amount of Positive and Negative Links

In [87]:
for idx, row in selected_bugs.iterrows():
    amount_tcs = 0  # amount of testcases of feature
    for f_id in row.Features_IDs.split(" "):
        if f_id != "":
            amount_tcs = amount_tcs + len(testcases[testcases.Feature_ID == int(f_id)])

    selected_bugs.at[idx, 'Amount_TCs'] = amount_tcs
    
display(selected_bugs[['Bug_Number', 'Features_IDs', 'Amount_TCs']].head(10))

positives_amount = selected_bugs.Amount_TCs.sum()
negatives_amount = len(selected_bugs) * len(testcases) - positives_amount

print("Total Amount of Expected Positive Links: {}".format(positives_amount))
print("Total Amount of Expected Negative Links: {}".format(negatives_amount))

/home/guilherme/anaconda3/envs/trace-link-recovery-study/lib/python3.6/site-packages/pandas/core/indexing.py:362: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self.obj[key] = _infer_fill_value(value)
/home/guilherme/anaconda3/envs/trace-link-recovery-study/lib/python3.6/site-packages/pandas/core/indexing.py:543: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self.obj[item] = s


Bug_Number Features_IDs  Amount_TCs
9192      1181835            6        31.0
10621     1248267                      0.0
10622     1248268                      0.0
10904     1257087            1        13.0
11150     1264988                      0.0
11229     1267480            3        22.0
11232     1267501                      0.0
11331     1269348            3        22.0
11339     1269485                      0.0
11375     1270274            6        31.0

Total Amount of Expected Positive Links: 915.0
Total Amount of Expected Negative Links: 18336.0


### ---------